### PR TITLE
Fix misaligned QoS settings

### PR DESCRIPTION
**Pull request by ctron**
This change aligns the consumer QoS with the producer QoS for telemetry data.

### CODE REVIEW

**Comment by ctron** [comment]
@grs @dejanb @gemmellr are the ones who can explain this. Big thanks!

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractConsumer.java [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractConsumer.java]: thanks for spotting

**Comment by ctron** [comment]
> What is the growing backlog issue?

I did see in the router statistics a growing number of "unsettled" messages. Aligning those two settings fixed this as (IIRC) @grs suggested. Although I don't have any understanding of "why".

**Comment by dejanb** [comment]
@sophokles73 there's a bug we discovered during enmasse testing, when sender is using unsettled and consumer pre-settled semantics. It's tracked by

https://issues.apache.org/jira/browse/DISPATCH-882

But no matter that, I think consumer should use AT_LEAST_ONCE after the changes in the messaging service.

**Comment by sophokles73** [comment]
Hmm, the issue does not (yet) have a classification as a bug, i.e. currently it is unclear what the dispatch router's expected behavior in this case should be. I have added a comment expressing my expectation, though. However, the change proposed in this PR would also require us to change the Telemetry API accordingly. At the moment, it reads:
> Hono supports *AT MOST ONCE* delivery of messages only. A client therefore MUST use `settled` for the *snd-settle-mode* and `first` for the *rcv-settle-mode* fields of its attach frame during link establishment. All other combinations are not supported by Hono and result in the termination of the link.

We could simply *relax* this requirement by also allowing *AT_LEAST_ONCE* delivery in the API spec and having our own client use this mode (as proposed by this PR).  While using *AT_MOST_ONCE* for delivering downstream telemetry messages seemed to be a good idea at the time of the API's inception, I now think that it actually has no benefit compared to using *AT_LEAST_ONCE*. With *AT_MOST_ONCE* we might end up with messages piling up in the dispatch router if the consumers are too slow to process them at the same rate as they come in. Even if the router is fixed in a way that it simply discards excess messages in this case (instead of buffering them in memory), we would still have all those excess messages having been accepted by the protocol adapters and flown through the system. On the other hand, if we used *AT_LEAST_ONCE* and we had slow consumers, the dispatch router would start to block the upstream producers once it runs out of credit, thus effectively propagating the back pressure upstream to the protocol adapters which would then discard the (excess) messages when they are published by the devices, i.e. these messages would not get sent downstream at all. I guess this is actually what we wanted to achieve in the first place by employing AMQP's flow control mechanism :-) Or am I missing something?

**Comment by sophokles73** [comment]
@pellmann @ppatierno @dejanb @sysexcontrol 
WDYT?

BTW the bug referred to by Dejan doesn't seem to be related to a general inability to forward presettled messages to unsettled receivers. Anyway, based on my argumentation above I still think that it would be the right thing to change the Telemetry API accordingly and only use *AT_LEAST_ONCE* (unsettled) delivery.

**Comment by pellmann** [comment]
For telemetry I think the most important feature is, that we burden our messaging infrastructure as less as possible. This means:

1. have full flow control, so we can discard messages as soon as possible, if the consumer can not process them (at the device or at the adapter).
2. do not buffer or process messages in the system, that will not made it to the consumer instantly 

Since this seems to be only possible with AT_LEAST_ONCE at the consumer link, we should change this accordingly.

**Comment by sysexcontrol** [comment]
Although it needs some thoughts why this makes sense, I agree to changing this to AT_LEAST_ONCE as well.

What made me hesitating shortly was that there will be acknowledgements for every single telemetry message which (usually) are not of interest to the protocol adapter but probably are just discarded.
But since this seems to be the only way to enable proper flow control, this should not be significant IMHO.

So: +1 from my side.

**Comment by dejanb** [comment]
While I don't have nothing against switching to AT_LEAST_ONCE in general, I'm not sure we're gonna achieve what we want with the change. What should be the point of discarding messages early in this case? If the system have enough throughput to send those messages downstream, but consumers are the bottleneck, then there's no problem in discarding them on the router? If the throughput of the Hono is the problem, then I'm not sure if sending settlements for every message will help us?
Anyhow, I think if make the change we should go all the way through? Also, what happens with events API in this case. What will be the difference between telemetry and events if they both use the same QoS. Do we need to document that events are gonna be brokered?

**Comment by sophokles73** [comment]
> What should be the point of discarding messages early in this case? If the system have enough throughput to send those messages downstream, but consumers are the bottleneck, then there's no problem in discarding them on the router?

FMPOV It is exactly the other way around: if there are no consumers for the data, then why should Hono's protocol adapters accept it from devices and burn up CPU cycles and (inter-service) bandwidth?

> Also, what happens with events API in this case. What will be the difference between telemetry and events if they both use the same QoS. Do we need to document that events are gonna be brokered?

The difference is that for events the protocol adapters wait for the  *accepted* outcome and only then indicate to the device that the data has been accepted whereas for telemetry data the protocol adapters do *not* wait for the messages being settled and immediately indicate to devices that they have accepted the data.

**Comment by dejanb** [comment]
> FMPOV It is exactly the other way around: if there are no consumers for the data, then why should Hono's protocol adapters accept it from devices and burn up CPU cycles and (inter-service) bandwidth?

Yes, but when we actually have a load of data to transmit we will burn more resources using "at least once" (cpu and bandwidth) and thus lower the throughput of the system when we need it the most. 

> The difference is that for events the protocol adapters wait for the accepted outcome and only then indicate to the device that the data has been accepted whereas for telemetry data the protocol adapters do not wait for the messages being settled and immediately indicate to devices that they have accepted the data.

OK, that makes sense. Thanks!

**Comment by sophokles73** [comment]
> Yes, but when we actually have a load of data to transmit we will burn more resources using "at least once" (cpu and bandwidth) and thus lower the throughput of the system when we need it the most.

Why is that? The overhead of sending a (very small) disposition frame upstream for every message that we actually want to process should in any case outweigh the overhead of sending (potentially large) telemetry messages downstream that we cannot deliver to a consumer in the first place, don't you think? FMPOV we can save much more bandwidth by keeping non-deliverable telemetry messages out of the system than by preventing disposition frames being sent upstream.

**Comment by dejanb** [comment]
> Why is that? The overhead of sending a (very small) disposition frame upstream for every message that we actually want to process should in any case outweigh the overhead of sending (potentially large) telemetry messages downstream that we cannot deliver to a consumer in the first place, don't you think? FMPOV we can save much more bandwidth by keeping non-deliverable telemetry messages out of the system than by preventing disposition frames being sent upstream.

We are clearly discussing all this theoretically, looking at different use cases? The answer is: it depends; on how many messages we are handling and if we are rejecting and how many and how big messages.

One idea that occurred to me is that we could leave consumers using AT_MOST_ONCE semantics (after a router fix, this should work with AT_LEAST_ONCE senders correctly) and make sender semantics configurable (default AT_LEAST_ONCE). In that way we can have a flow control by default, but allow people to turn it off for max throughput if they are sure that they can have proper consuming capacity. 

And we can actually at some point run tests and verify all this for various use cases and deployments.

WDYT?

**Comment by sophokles73** [comment]
> One idea that occurred to me is that we could leave consumers using AT_MOST_ONCE semantics (after a router fix, this should work with AT_LEAST_ONCE senders correctly) and make sender semantics configurable (default AT_LEAST_ONCE). In that way we can have a flow control by default, but allow people to turn it off for max throughput if they are sure that they can have proper consuming capacity.

I would rather not re-start the discussion about using unsettled messages for sending telemetry messages downstream again. We made that decision to *always* use unsettled messages in order to get proper flow control between protocol adapters and downstream consumers.

I share @dejanb 's hope that consumers using presettled messages (*AT_MOST_ONCE*) should work once the bug in Dispatch Router has been fixed. I do not see, why using presettled messages for sending messages downstream should provide *max throughput* comparing to unsettled messages. The bandwidth for sending the *transfer* frames downstream is not eaten up by the *disposition* frames being sent upstream. Since the protocol adapters do not wait for the disposition frames when sending telemetry data, I do not see the point of using presettled messages at all.

**Comment by sophokles73** [comment]
It seems that we have a hard time coming to a conclusion here :-) Since this is about changing the Telemetry API in a non-compatible way, I would like to have a formal vote on the following options:

1. We change the Telemetry API to mandate *AT LEAST ONCE* delivery for consuming Telemetry data.
2. We change the Telemetry API to allow either *AT MOST ONCE* or *AT LEAST ONCE* delivery for consuming Telemetry data (as we already do for uploading Telemetry data) and *recommend* usage of *AT LEAST ONCE* delivery to better support end-to-end flow control.
3. We leave the Telemetry API as is and allow *AT MOST ONCE* delivery only.

@pellmann @dejanb @sysexcontrol @ppatierno 

please cast your vote in this issue ASAP so that we can move on with this and can have it fixed for the imminent 0.5 release.

BTW my vote is for option 2.

**Comment by sysexcontrol** [comment]
Considering the details of the discussion that took place before, I vote for option 2 as well.

We can find out later or for special cases if the (unnecessary) *disposition* frames being sent upstream account for any relevant CPU usages or if they really just don't matter. 
And even if they should matter: to me good flow control is essential for production setups, something that most times is ignored by self-written connectivity (which then break down in heavy load scenarios and just declare it "out of scope" ;-) ). Thus option 3 is completely out IMHO.

So, :+1:  for option 2.

**Comment by pellmann** [comment]
I share @dejanb view, that the discussion is very theoretically, because we do not have much experience with our system in the real world. And from Hono's point of view there is some AMQP network and it seems that the handling of the delivery guarantees might be very different in different implementations. So it is a good idea to allow different settlement strategies at the API level.

Vote for option 2.

**Comment by ppatierno** [comment]
I think that having only AT_MOST_ONCE for the consumer is a limitation. During sessions (at conferences) I had speaking about Hono different people asked me having a more "reliable" delivery avoiding just dropping telemetry data instead of stopping the producer if consumer can't keep the pace.
I think that option 2 makes sense. +1 from me.

**Comment by sophokles73** [comment]
> During sessions (at conferences) I had speaking about Hono different people asked me having a more "reliable" delivery avoiding just dropping telemetry data instead of stopping the producer if consumer can't keep the pace.

@ppatierno, make no mistake here, allowing for *AT LEAST ONCE* delivery for telemetry does **not** increase the QoS! It only means that the messages will be discarded by the protocol adapters already if not enough downstream capacity is available. But the messages will still be dropped. However, the HTTP adapter will be able to indicate this fact to the device immediately.

**Comment by ppatierno** [comment]
@sophokles73 you are absolutely right sorry ! So request from people at conferences is still opened ... but don't need to be addressed :-)

**Comment by sophokles73** [comment]
Based on the feedback received so far, I will merge this PR and will go ahead adapting the Telemetry API spec accordingly.

**Pull request merged**